Added validation for the context of extensions

Kasdejong opened the pull request:
## Description
The validator should now (finally) validate extension contexts against their expected context.

## Related issues
Closes #267 

## Tests
Added tests for all possible extension contexts

Kasdejong committed:
wip

Kasdejong committed:
Merge branch 'develop' into feature/extension-context-validation

# Conflicts:
#	src/Firely.Fhir.Validation/Impl/FhirPathValidator.cs
#	test/Firely.Fhir.Validation.Compilation.Tests.Shared/FhirTestCases

Kasdejong committed:
Added tests and finalized implementation for extension context validation

Kasdejong committed:
wip

Kasdejong committed:
a bit better

Kasdejong committed:
more progress, only the slices remain

Kasdejong committed:
correct link

Kasdejong committed:
hopefully fixed slicing

Kasdejong committed:
forgot to commit

Kasdejong commented:
We have an issue here (Apart from the fact that I seem to have done something weird with the subrepo). We do not step into every profile we encounter. This makes sense for some cases, except it makes verifying whether we are in the correct Element a lot harder, and as a bonus, I believe omitting it is a wrong assumption.

In a separate branch, I created a Patient with a link (backboneElement) with multiple id's and, lo and behold, the validation is a success! I believe us not invoking the profile of backboneElement is a mistake here. This would require quite a big change to the validator though, but it would make my work a LOT easier. I will leave this branch for now until we have discussed this internally.

Kasdejong committed:
added backboneElement refs to schema snaps

Kasdejong committed:
getting closer...

Kasdejong committed:
it works now!

Kasdejong committed:
we ended up not needing these IDs

Kasdejong committed:
changed these back

Kasdejong committed:
Merge branch 'develop' into feature/extension-context-validation

# Conflicts:
#	test/Firely.Fhir.Validation.Compilation.Tests.Shared/FhirTestCases

Kasdejong committed:
it seems as though this is more correct. Changed implementation of Extension-Context.Resource

ewoutkramer commented on .idea/.idea.Firely.Validator.API/.idea/projectSettingsUpdater.xml:
```diff
@@ -1,6 +1,6 @@
 <project version="4">
   <component name="RiderProjectSettingsUpdater">
-    <option name="vcsConfiguration" value="2" />
+    <option name="vcsConfiguration" value="3" />
```
We should Git .ignore this file

ewoutkramer commented on src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/ExtensionContextBuilder.cs:
```diff
@@ -0,0 +1,22 @@
+/// <summary>
+/// 
+/// </summary>
+#pragma warning disable CS0618 // Type or member is obsolete
```
Do we need this pragma?

Kasdejong replied:
I'm afraid there are so many obsolete classes here that we do

ewoutkramer commented on src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs:
```diff
@@ -72,6 +72,13 @@ public IEnumerable<IAssertion> Build(ElementDefinitionNavigator nav, ElementConv
             }
+            else
+            {
+                if (def.Type.SingleOrDefault()?.Code is { } tc)
```
This is very tricky code, so please add so comments on what and why we are doing this.

Kasdejong committed:
should be last commit, finally!

Kasdejong committed:
should be last commit, finally!

Kasdejong committed:
Pushed to wrong branch

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/ExtensionContextValidator.cs:
```diff
@@ -0,0 +1,182 @@
+        Invariants = invariants.ToList();
+    }
+
+    internal List<TypedContext> Contexts { get; }
```
These members must be serializable (which I think they are), and marked as `[DataMember]` and use `IReadOnlyCollection<>`, see `ElementSchema.cs` for an example.

Note: they must be serializable since another project of ours (the CAR factory) serializes the element schemas to a sqlite database.

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/FhirPathValidator.cs:
```diff
@@ -105,22 +104,44 @@ public override JToken ToJson()
+            }
+        }
+        
+        internal InvariantResult RunInvariant(ScopedNode input, ValidationSettings vc, ValidationState s, params (string key, IEnumerable<ITypedElement> value)[] env)
```
Is there a way to avoid the copy & paste?  It seems only the environment differs....

Kasdejong committed:
this check is necessary, otherwise it could crash on an extension without 2 parents

Kasdejong committed:
removed some unnecessary code

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/SliceValidator.cs:
```diff
@@ -40,6 +40,11 @@ public class SliceValidator : IGroupValidatable
+            /// <summary>
+            /// Identifier for the slice.
+            /// </summary>
+            public string? Id { get; private set; }
```
This needs `[DataMember]` too.

ewoutkramer replied:
How Is this different from the name?

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/SliceValidator.cs:
```diff
@@ -72,12 +77,12 @@ public SliceCase(string name, IAssertion condition, IAssertion? assertion)
 
             /// <inheritdoc cref="IJsonSerializable.ToJson"/>
-            public JToken ToJson() =>
+            public JToken ToJson() => 
```
Shouldn't the new `id` be in here too?

Kasdejong replied:
Turns out Id is unused. I simply forgot to remove the property.

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/SliceValidator.cs:
```diff
@@ -246,12 +251,22 @@ public void AddToSlice(SliceCase slice, IScopedNode item, int originalIndex)
+            {
+                // we check for any slices' type, and use that as the type for the whole slice
+                var type = this.Keys
+                    .Select(sliceCase => sliceCase.Assertion)
```
Just checking the first assertion should do, right?

Kasdejong replied:
The first assertion is not always an elementschema as far as I'm aware. In fact, I've seen "unknown type" get triggered at times. I'm not sure what happens here.

ewoutkramer replied:
Ok, we should understand this before putting it into production, so let's go over it one more time.

Kasdejong committed:
how does this change every time?

Kasdejong committed:
oops...

ewoutkramer commented on test/Firely.Fhir.Validation.Compilation.Tests.R4/SchemaSnaps/Bundle.json:
```diff
@@ -191,6 +191,7 @@
         "humanDescription": "must be a resource unless there's a request or response"
       },
       "cardinality": "0..*",
+      "baseRef": "BackboneElement",
```
This is not a ref, so maybe "baseType" ?

ewoutkramer commented on test/Firely.Fhir.Validation.Compilation.Tests.R4/SchemaSnaps/ProfiledObservation.json:
```diff
@@ -820,6 +820,7 @@
         "humanDescription": "Must have at least a low or a high or text"
       },
       "cardinality": "0..*",
+      "baseRef": "BackboneElement",
```
Shouldn't the slices have a baseRef too?

ewoutkramer commented on test/Firely.Fhir.Validation.Compilation.Tests.Shared/ValidationIntegrationTests.cs:
```diff
@@ -52,7 +52,7 @@ private static string getSuiteDirectory(string suiteName) =>
         public async Task RunValidateTestSuite(string suiteName)
         {
-            var overwrite = false;
+            var overwrite = true;
```
This should be returned to false.

ewoutkramer commented on src/Firely.Fhir.Validation/Schema/DefinitionPath.cs:
```diff
@@ -40,15 +45,65 @@ public bool HasDefinitionChoiceInformation
 
+        internal bool MatchesContext(string contextEid)
+        {
+            var match = new Regex("^(?<type>[A-Za-z]*).?(?<location>.*)$").Match(contextEid);
```
For regexes that we use unaltered multiple times, you'd better cache the compiled regex, see https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-regex#static-regular-expressions

ewoutkramer commented on src/Firely.Fhir.Validation/Schema/DefinitionPath.cs:
```diff
@@ -40,15 +45,65 @@ public bool HasDefinitionChoiceInformation
+            
+            var type = match.Groups["type"].Value;
+            var location = match.Groups["location"].Value;
+            var locationComponents = location == "" ? [] : location.Split('.', ':');
```
`location.Split('.',':', StringSplitOptions.RemoveEmptyEntries)` will do the same without the empty check. If you want to do an empty check, use `String.Empty` instead of `""` to avoid creating empty strings.

ewoutkramer commented on src/Firely.Fhir.Validation/Schema/DefinitionPath.cs:
```diff
@@ -40,15 +45,65 @@ public bool HasDefinitionChoiceInformation
+
+            return current switch
+            {
+                ChildNavEvent cne => cne.ChildName == location.Last() && matchesContext(current.Previous, type, location[..^1]),
```
Maybe `Span<T>` works more efficiently here in these recursive calls?

ewoutkramer commented on src/Firely.Fhir.Validation/Schema/DefinitionPath.cs:
```diff
@@ -40,15 +45,65 @@ public bool HasDefinitionChoiceInformation
+            if (location.Length == 0)
+                return current switch
+                {
+                    ChildNavEvent cne => cne.Type == type || type == "Element",
```
Can you give concrete examples of the extension paths you are matching? This will make this easier to read in a few years or so ;-)

ewoutkramer commented on src/Firely.Fhir.Validation/Schema/PathStack.cs:
```diff
@@ -62,17 +62,22 @@ internal abstract class PathStackEvent
 
         public string ChildName { get; }
         public string? ChoiceType { get; }
+        public string? Type { get; }
```
Maybe `internal` because we don't print it either.

ewoutkramer commented on src/Firely.Fhir.Validation/Schema/PathStack.cs:
```diff
@@ -158,12 +163,14 @@ protected internal override string Render()
         }
 
         public string SliceName { get; }
+        public string Type { get; }
```
Idem

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/ExtensionContextValidator.cs:
```diff
@@ -0,0 +1,176 @@
+        if (Contexts.Any(c => c.Type == null))
+        {
+            return new IssueAssertion(Issue.PROFILE_ELEMENTDEF_INCORRECT,
+                    "Extension context type was not set, but a context was defined. Skipping non-invariant context validation")
```
Shouldn't we report this in the constructor? Then we can catch it before we ship the validation to the client.

Kasdejong replied:
We can still validate invariants. What should we do here?

ewoutkramer replied:
I design principle for the validator is that by the time we are validating, we do not report errors anymore about the quality of the profile itself, otherwise put, we cannot bother the end user (who is merely expecting issues about his/her/their instance) with errors that need to be corrected by the author of the profile.  So, we should report this error during compilation.

What we could do (but I don't know that we did it before), is give a WARNING during compilation (which I don't think we can yet), that without this set, only half of the validation will be done. But for now, we should be strict. An incorrect extension context should be flagged as far as I am concerned.

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/ExtensionContextValidator.cs:
```diff
@@ -0,0 +1,176 @@
+        if (failedInvariantResults.Count != 0)
+        {
+            return ResultReport.Combine(
+                failedInvariantResults
```
If you don't want to repeat the `fr.Report is not null`, you can also group by `fr.Report is not null` ;-).  You'll get a group `true` and `false` and then you can run the different behaviours, instead of the concat.

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/ExtensionContextValidator.cs:
```diff
@@ -0,0 +1,176 @@
+        return context.Type switch
+        {
+            ContextType.DATATYPE => contextNode.InstanceType == context.Expression,
+            ContextType.EXTENSION => contextNode.InstanceType == "Extension" && (contextNode.Parent?.Children("url").SingleOrDefault()?.Value as string ?? "None") == context.Expression,
```
Maybe use a more random magic string (null would also not match btw).

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/ExtensionContextValidator.cs:
```diff
@@ -0,0 +1,176 @@
+
+    private static InvariantValidator.InvariantResult runContextInvariant(IScopedNode input, string invariant, ValidationSettings vc, ValidationState state)
+    {
+        var fhirPathValidator = new FhirPathValidator("ctx-inv", invariant.Replace("%extension", "%%extension"));
```
Document that we use `%%extension` in our fhirpath dialect, so we need to replace this.

ewoutkramer replied:
Or do we need to change this in our FP implementation?

ewoutkramer replied:
I think we should. For 5.x we should then accept both. But it will; fix this.

Kasdejong replied:
https://github.com/FirelyTeam/firely-net-sdk/issues/2911

Kasdejong committed:
applied requested changes

Kasdejong committed:
public api changes

Kasdejong committed:
changed implementation for extension type extension context because it was inconsistent. Also updated manifest.

Kasdejong committed:
fixed no invariant result showing when the invariant evaluated to false

ewoutkramer commented on src/Firely.Fhir.Validation/Impl/baseType.cs:
```diff
@@ -0,0 +1,35 @@
+#else
+[System.Obsolete("This function is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.")]
+#endif
+public class baseType : IAssertion
```
Every type name should start with a captial letter.

Kasdejong committed:
PR changes (mainly type names)

Kasdejong committed:
removed TypedContext api warnings

ewoutkramer committed:
Ran code fix for public API checker

ewoutkramer committed:
Rename baseType.cs to BaseType.cs

ewoutkramer committed:
Update BaseType.cs

Since BaseType needs to be serializable, I had to add [DataMember] to its only property.
